### PR TITLE
Add IEC62056 (DLMS/COSEM) protocol dissector

### PR DESCRIPTION
**Pull request by 0xA50C1A1**
Please sign (check) the below before submitting the Pull Request:

- [x] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [x] I have read the contributing guide lines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [x] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Describe changes:

DLMS is a protocol used to communicate with water, gas and electricity meters.

### CODE REVIEW

**Comment by 0xA50C1A1** [comment]
Looks like that JA4 related commits broke the build.

File: src/lib/ndpi_analyze.c [hunk 1]
```diff
@@ -1765,6 +1765,15 @@ u_int16_t ndpi_crc16_xmodem(const void *data, size_t n_bytes) {
...
 
+u_int16_t ndpi_crc16_x25(const void* data, size_t n_bytes) {
+  u_int16_t crc = 0xFFFF;
+  u_int8_t* b = (u_int8_t*)data;
+  while (n_bytes--) {
+    crc = (crc >> 8) ^ crc16_ccitt_table[(crc ^ *b++) & 0xFF];
+  }
+  return (crc ^ 0xFFFF);
+}
+
```
Comment (IvanNardi) [inline src/lib/ndpi_analyze.c:1841]: Could you add this new function to `fuzz/fuzz_alg_crc32_md5.c`, please?

File: tests/cfgs/default/result/dlms.pcapng.out [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+Patricia risk mask:   0/0 (search/found)
+Patricia risk mask IPv6: 0/0 (search/found)
+Patricia risk:        0/0 (search/found)
+Patricia risk IPv6:   0/0 (search/found)
+Patricia protocols:   2/0 (search/found)
+Patricia protocols IPv6: 0/0 (search/found)
+
+IEC62056	284	25819	1
+
+	1	TCP 192.168.1.4:55214 <-> 192.168.1.12:4060 [proto: 378/IEC62056][IP: 0/Unknown][ClearText][Confidence: DPI][DPI packets: 4][cat: IoT-Scada/31][188 pkts/14674 bytes <-> 96 pkts/11145 bytes][Goodput ratio: 15/43][0.26 sec][bytes ratio: 0.137 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 1/3 29/32 2/4][Pkt Len c2s/s2c min/avg/max/stddev: 66/54 78/116 183/194 15/46][Risk: ** Known Proto on Non Std Port **][Risk Score: 50][Plen Bins: 76,8,0,2,13,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
```
Comment (IvanNardi) [inline tests/cfgs/default/result/dlms.pcapng.out]: Do you have also an UDP example?

**Review by IvanNardi — CHANGES_REQUESTED** [review]

File: tests/cfgs/default/result/dlms.pcapng.out (hunk 2 above)
Comment (0xA50C1A1) [reply tests/cfgs/default/result/dlms.pcapng.out]: > Do you have also an UDP example?

Well, I found a good one.

File: src/lib/ndpi_analyze.c (hunk 1 above)
Comment (0xA50C1A1) [reply src/lib/ndpi_analyze.c:1841]: > Could you add this new function to `fuzz/fuzz_alg_crc32_md5.c`, please?

Sure.

**Comment by 0xA50C1A1** [comment]
@IvanNardi Thanks for the review, I've fixed it.

**Comment by IvanNardi** [comment]
@0xA50C1A1, could you confirm that Wireshark is not able to recognize this protocol?

**Comment by 0xA50C1A1** [comment]
> @0xA50C1A1, could you confirm that Wireshark is not able to recognize this protocol?

Unfortunately, there is still no plugin for this protocol in Wireshark. Well, there was one [out-of-tree plugin](https://github.com/bearxiong99/wireshark-dlms), but its development has long been abandoned. There's also a [Lua script](https://github.com/matousp/dlms-analysis/tree/master). By the way, I got that pcap samples from these two repositories.

**Comment by IvanNardi** [comment]
> > @0xA50C1A1, could you confirm that Wireshark is not able to recognize this protocol?
> 
> Unfortunately, there is still no plugin for this protocol in Wireshark. Well, there was one [out-of-tree plugin](https://github.com/bearxiong99/wireshark-dlms), but its development has long been abandoned. There's also a [Lua script](https://github.com/matousp/dlms-analysis/tree/master). By the way, I got that pcap samples from these two repositories.

Thanks for this information. I added it to `doc/protocols.rst`

**Review by IvanNardi — APPROVED** [review]
I am learning a lot of interesting uncommon stuff these days... :)
Thanks!

**Pull request merged**